### PR TITLE
Added recipe for kwant

### PR DESCRIPTION
**Pull request by basnijholt**
Depends on the following pull requests:
- [x] `tinyarray` #1444
- [x] `metis` #1442
- [x] `mumps` #1445
- [x] `scotch` #1427

### CODE REVIEW

**Comment by basnijholt** [comment]
@jakirkham could you give me advice on the following part:
```
requirements:
  build:
    - gcc
    - openblas  # [linux]
  run:
    - openblas  # [linux]
    - libgcc
    - libgfortran  # [linux]
```

Should I swap `gcc` with `toolchain` and omit `libgcc`? And what about `openblas` and gfortran`?

**Comment by basnijholt** [comment]
> For **recipes/kwant**:
> * Failed to even lint the recipe (might be a conda-smithy bug) 😢

@pelson @msarahan any idea why this can have happened? [This](https://github.com/basnijholt/staged-recipes/blob/kwant/recipes/kwant/meta.yaml) is the `meta.yaml`

File: recipes/kwant/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,73 @@
+{% set name = "kwant" %}
+
+package:
+  name: {{ name }}
+  version: "1.2.2" [py3k]
```
Comment (ocefpaf) [inline recipes/kwant/meta.yaml]: The linter is choking here. You can lint a recipe locally with `conda smithy recipe-lint`.

File: recipes/kwant/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  version: "1.1.2" [py27]
+
+source:
+  git_url: https://gitlab.kwant-project.org/kwant/kwant.git
+  git_tag: v1.2.2 [py3k]
+  git_tag: v1.1.2 [py27]
```
Comment (ocefpaf) [inline recipes/kwant/meta.yaml]: @basnijholt change the top of your recipe to

```yaml
{% set name = "kwant" %}
{% set py3k_ver = "1.2.2" %}
{% set py2k_ver = "1.1.2" %}

package:
  name: {{ name }}
  version: {{ py3k_ver }}  # [py3k]
  version: {{ py2k_ver }}  # [py2k]

source:
  fn: v{{ version }}.tar.gz
  url: https://gitlab.kwant-project.org/kwant/kwant/archive/v{{ version }}.tar.gz
  sha256: 8c20230f266622e5fa4c75fd3b1e8018d9a99643e71b86a7e91cb7b6902d8620  # [py3k]
  sha256: 3229009580e33e58f5cfb8405c44f8146b89b621b6ebe599166ef9add8ec9755  # [py2k]
```

That also fixed the linter issue locally for me.  BTW do the different version reflect the Python versions? If so maybe adding a `py3k` version first and than a maintenance branch might be a better strategy.

Comment (basnijholt) [inline recipes/kwant/meta.yaml]: Thanks a lot! This fixed the issue :)

File: recipes/kwant/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+build:
+  skip: true  # [win]
+  number: 0
+  binary_relocation: true
+
+requirements:
+  build:
+    - python
+    - gcc
```
Comment (jakirkham) [inline recipes/kwant/meta.yaml]: Please use `toolchain` instead.

File: recipes/kwant/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    - python
+    - openblas  # [linux]
+    - mumps_seq
+    - numpy
+    - matplotlib
+    - scipy
+    - cython
+    - nose
+    - tinyarray
+    - libgcc
```
Comment (jakirkham) [inline recipes/kwant/meta.yaml]: Please drop this as it is unneeded.

File: recipes/kwant/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    - libgfortran  # [linux]
```
Comment (jakirkham) [inline recipes/kwant/meta.yaml]: Please drop this as it should also be unneeded.

**Comment by jakirkham** [comment]
FWICT this is Python, C, and C++. However there is not Fortran that I can see. The following suggestions should be fine.

Also does this link to a BLAS directly? If so, mimic our NumPy recipe w.r.t. OpenBLAS.

**Comment by basnijholt** [comment]
There is definitely Fortran there, also the package on [my channel](https://anaconda.org/basnijholt/kwant) doesn't work without `gfortran`. Are there differences for packages compiled by myself and `conda-forge`?

**Comment by basnijholt** [comment]
> Also does this link to a BLAS directly? If so, mimic our NumPy recipe w.r.t. OpenBLAS.

Yes, `kwant` directly links to BLAS, because it uses `LAPACK`.

**Comment by basnijholt** [comment]
@jakirkham you mean that I should add:
```
  features:
    - blas_openblas  # [linux]
```
as I did [here](https://github.com/basnijholt/staged-recipes/blob/812837ad668f4cf47ce55ab8172af288d7c11677/recipes/kwant/meta.yaml#L23)

However, in the [numpy recipe](https://github.com/conda-forge/numpy-feedstock/blob/master/recipe/meta.yaml#L26) I also see `- blas 1.1 openblas`. Don't know what it's for, but should I include that as well?

Another thing, why is OpenBLAS also added for Mac OSX, I thought this wasn't needed?

**Comment by basnijholt** [comment]
Builds succeed, are we ready to merge?

File: recipes/kwant/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    - gcc
+    - blas 1.1 {{ variant }}
+    - openblas  0.2.18|0.2.18.*
+    - mumps
+    - numpy
+    - matplotlib
+    - scipy
+    - cython
+    - nose
+    - tinyarray
```
Comment (jakirkham) [inline recipes/kwant/meta.yaml]: Are all of these needed at build time or can some of them be dropped (e.g. `matplotlib`, `scipy`, etc.)?

File: recipes/kwant/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    - kwant.linalg.tests
+    - kwant.physics
+    - kwant.physics.tests
+    - kwant.solvers
+    - kwant.solvers.tests
+    - kwant.tests
+
+about:
+  home: http://kwant-project.org/
+  license: BSD 2-Clause
```
Comment (jakirkham) [inline recipes/kwant/meta.yaml]: Could optionally add `license_family: BSD`.

**Comment by basnijholt** [comment]
> Are all of these needed at build time or can some of them be dropped (e.g. `matplotlib`, `scipy`, etc.)?

@jakirkham I have enumerated them all, and the current build requirements are all needed.

File: recipes/kwant/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,83 @@
+{% set name = "kwant" %}
+{% set py3k_ver = "1.2.2" %}
+{% set py2k_ver = "1.1.2" %}
+{% set variant = "openblas" %}
+
+package:
+  name: {{ name }}
+  version: {{ py3k_ver }}  # [py3k]
+  version: {{ py2k_ver }}  # [py2k]
```
Comment (jakirkham) [inline recipes/kwant/meta.yaml]: Why do we have 2 different versions? Could we not just start with 1.1.2 and then drop Python 2.x support when upgrading or is there more going on here?

File: recipes/kwant/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    - openblas 0.2.18|0.2.18.*
+    - mumps
+    - numpy
+    - matplotlib
+    - scipy
+    - cython
+    - nose
+    - tinyarray
+    - libgfortran
+    - libgcc
```
Comment (jakirkham) [inline recipes/kwant/meta.yaml]: I suspect we only need one of these. Maybe just `libgfortran`, but please let me know if there is more going on (e.g. is there C++?).

File: recipes/kwant/meta.yaml (hunk 7 above)
Comment (jakirkham) [inline recipes/kwant/meta.yaml]: Done below.

File: recipes/kwant/meta.yaml (hunk 9 above)
Comment (basnijholt) [inline recipes/kwant/meta.yaml]: They are both needed I think, see [this build](https://circleci.com/gh/conda-forge/staged-recipes/10748?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) of this commit 471c711

File: recipes/kwant/meta.yaml (hunk 8 above)
Comment (basnijholt) [inline recipes/kwant/meta.yaml]: Development for Python 2 stopped, however, bug fixes will still be applied. So the version number for both Python 2 and 3 will change, but not be the same.

File: recipes/kwant/meta.yaml (hunk 6 above)
Comment (basnijholt) [inline recipes/kwant/meta.yaml]: I removed matplotlib in 0e427a9

File: recipes/kwant/meta.yaml (hunk 9 above)
Comment (jakirkham) [inline recipes/kwant/meta.yaml]: I see. So it does need C++.

Please add `yum_requirements.txt` with `devtoolset-2-gcc-gfortran`. Also please add a selector to use `gcc` on OS X only. Then please add `export LIBRARY_PATH="${PREFIX}/lib"` to `build.sh`. Finally please drop `libgcc`, but keep/add `libgfortran` (in `build` and `run`).

File: recipes/kwant/meta.yaml (hunk 6 above)
Comment (jakirkham) [inline recipes/kwant/meta.yaml]: Yeah the comment moved. Seems like some GitHub bug with diffs. Thanks for doing that though.

File: recipes/kwant/meta.yaml (hunk 9 above)
Comment (jakirkham) [inline recipes/kwant/meta.yaml]: This strategy is used for `scipy`, `hdf5`, and other C/C++/Fortran libraries of note.

xref: https://github.com/conda-forge/scipy-feedstock/pull/2
xref: https://github.com/conda-forge/scipy-feedstock/pull/3

File: recipes/kwant/meta.yaml (hunk 8 above)
Comment (basnijholt) [inline recipes/kwant/meta.yaml]: Changed to only Python 2 in f32eb41.

File: recipes/kwant/meta.yaml (hunk 9 above)
Comment (basnijholt) [inline recipes/kwant/meta.yaml]: I did what you suggested, but still getting this error: `ImportError: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.15' not found (required by /opt/conda/envs/_test/lib/python2.7/site-packages/kwant/graph/slicer.so)`. I have added `export LD_LIBRARY_PATH="${PREFIX}/lib"` to `buid.sh`.

Comment (basnijholt) [inline recipes/kwant/meta.yaml]: Adding `LD_LIBRARY_PATH` alone didn't help, still the same error message.

File: recipes/kwant/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  version: {{ version }}
+
+source:
+  fn: v{{ version }}.tar.gz
+  url: https://gitlab.kwant-project.org/kwant/kwant/repository/archive.tar.gz?ref=v{{ version }}
+  sha256: a47544a8416b19a56da3d511610c9e09ed3bf84f66461d9f95f0f6fefebd0179
+
+build:
+  skip: true  # [win]
+  skip: true  # [py3k]
```
Comment (jakirkham) [inline recipes/kwant/meta.yaml]: These can be combined `skip: true  # [win or py3k]`. Also is it that this version is Python 2 only or is it that this version is the last one that supports Python 2, but also supports Python 3?

Comment (basnijholt) [inline recipes/kwant/meta.yaml]: Fixed in 82ea10cfd04e0355b19f591efa1106c2c19492b2. There is no version that combines both Python 2 and 3 simultaneously.

**Comment by jakirkham** [comment]
Also please add `[skip appveyor]` to commit messages as this does not build on Windows.

File: recipes/kwant/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+requirements:
+  build:
+    - python
+    - toolchain
+    - gcc  # [osx]
+    - blas 1.1 {{ variant }}
+    - openblas 0.2.18|0.2.18.*
+    - mumps
+    - numpy
```
Comment (basnijholt) [inline recipes/kwant/meta.yaml]: Yes it is linking to `numpy`. Fixed in e2bb0c2

**Comment by jakirkham** [comment]
So I think you have CircleCI enabled on your fork of staged-recipes. Could you please disable it? Seems to be getting in the way of running it here at staged-recipes.

**Comment by basnijholt** [comment]
I stopped it :-)

**Comment by jakirkham** [comment]
Toggling to get CircleCI to restart on staged-recipes ( hopefully 🍀 ).

**Comment by jakirkham** [comment]
Seems to have worked. 😄

**Comment by basnijholt** [comment]
And also the builds are passing 👍

**Comment by jakirkham** [comment]
Thanks @basnijholt .

**Pull request merged**